doc/shared/sunadjoint/SUNAdjointCheckpointScheme.rst

```diff
@@ -0,0 +1,273 @@
+   :param step_num: the step number of the checkpoint
+   :param stage_num: the stage number of the checkpoint
+   :param out: the loaded state vector
+
+   :return: A :c:type:`SUNErrCode` indicating failure or success.
+
+.. c:function:: SUNErrCode SUNAdjointCheckpointScheme_EnableDense(SUNAdjointCheckpointScheme cs, \
+   sunbooleantype on_or_off)
+
+   Enables or disables dense checkpointing (checkpointing every step/stage).
+
+   :param cs: The :c:type:`SUNAdjointCheckpointScheme` object
+   :param on_or_off: if true, dense checkpointing will be turned on, ifalse it will be turned off.
+
+   :return: A :c:type:`SUNErrCode` indicating failure or success.
+
+.. c:function:: SUNErrCode SUNAdjointCheckpointScheme_Destroy(SUNAdjointCheckpointScheme* cs_ptr)
+
+   Destroys (deallocates) the SUNAdjointCheckpointScheme object.
+
+   :param cs_ptr: pointer to a :c:type:`SUNAdjointCheckpointScheme` object
+
+   :return: A :c:type:`SUNErrCode` indicating failure or success.
+
+
+.. _SUNAdjointCheckpointScheme.Basic:
+
+The SUNAdjointCheckpointScheme_Basic Module
+===========================================
+
+The SUNAdjointCheckpointScheme_Basic module implements a scheme where a checkpoint is saved at some
+fixed interval (in timesteps). The module supports checkpointing of time step states only, or time
+step stages with intermediate stage states as well (for multistage methods). When used with a
+fixed timestep size then the number of checkpoints that will be saved is fixed. However, with
+adaptive timesteps the number of checkpoints stored with this scheme is unbounded.
+
+The diagram below illustrates how checkpoints are stored with this scheme:
+
+
+
+The SUNAdjointCheckpointScheme_Basic module has the following user-callable functions:
+
+.. c:function:: SUNErrCode SUNAdjointCheckpointScheme_Create_Basic(SUNDataIOMode io_mode, SUNMemoryHelper mem_helper, int64_t interval, int64_t estimate, sunbooleantype save_stages, sunbooleantype keep, SUNContext sunctx, SUNAdjointCheckpointScheme* check_scheme_ptr)
+
+   Creates a new `SUNAdjointCheckpointScheme` object that checkpoints at a fixed interval.
+
+   :param io_mode: The IO mode used for storing the checkpoints.
+   :param mem_helper: Memory helper for managing memory.
+   :param interval: The interval (in steps) between checkpoints.
+   :param estimate: An estimate of the total number of checkpoints needed.
+   :param save_stages: If using a multistage method, should stages be saved with the step.
+   :param keep: Keep data stored even after it is not needed anymore.
+   :param sunctx: The SUNContext for the simulation.
+   :param check_scheme_ptr: Pointer to the newly constructed object.
+   :return: A `SUNErrCode` indicating success or failure.
+
+.. c:function:: SUNErrCode SUNAdjointCheckpointScheme_ShouldWeSave_Basic(SUNAdjointCheckpointScheme check_scheme, sunindextype step_num, sunindextype stage_num, sunrealtype t, sunbooleantype* yes_or_no)
```

Comment:
I don't see why this and the function below need to be publicly documented. I haven't gotten to the header files, but I don't think they should be in the public ones. If I understand correctly, users would uses the base class functions like `SUNAdjointCheckpointScheme_ShouldWeSave` rather than `SUNAdjointCheckpointScheme_ShouldWeSave_Basic`

Reply:
>  If I understand correctly, users would uses the base class functions like `SUNAdjointCheckpointScheme_ShouldWeSave` rather than `SUNAdjointCheckpointScheme_ShouldWeSave_Basic`

That is correct, although we are inconsistent in the documentation w.r.t. documenting the implementation specific functions or not.